packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,786 @@
+    };
+  }
+
+  addBlock({blockRootHex, block, source}: AddBlock<ForkBlobsDA>): void {
+    if (!this.state.hasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INVALID_CONSTRUCTION,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot addBlock to BlockInputBlobs after it already has a block"
+      );
+    }
+
+    // this check suffices for checking slot, parentRoot, and forkName
+    if (blockRootHex !== this.blockRootHex) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISMATCHED_ROOT_HEX,
+          blockInputRoot: this.blockRootHex,
+          mismatchedRoot: blockRootHex,
+          source: source.source,
+          peerId: `${source.peerIdStr}`,
+        },
+        "addBlock blockRootHex does not match BlockInput.blockRootHex"
+      );
+    }
+
+    for (const {blobSidecar} of this.blobsCache.values()) {
+      if (!blockAndBlobArePaired(block, blobSidecar)) {
+        this.blobsCache.delete(blobSidecar.index);
+        // TODO: (@matthewkeil) spec says to ignore invalid blobs but should we downscore the peer maybe?
+        // this.logger?.error(`Removing blobIndex=${blobSidecar.index} from BlockInput`, {}, err);
+      }
+    }
+
+    const hasAllData = this.blobsCache.size === block.message.body.blobKzgCommitments.length;
+
+    this.state = {
+      ...this.state,
+      hasAllData,
+      block,
+      versionHashes: getVersionHashes(block),
+      source,
+      timeCompleteSec: hasAllData ? source.seenTimestampSec : undefined,
+    } as BlockInputBlobsState;
+    this.blockPromise.resolve(block);
+    if (hasAllData) {
+      this.dataPromise.resolve(this.getBlobs());
+    }
+  }
+
+  hasBlob(blobIndex: BlobIndex): boolean {
+    return this.blobsCache.has(blobIndex);
+  }
+
+  addBlob({blockRootHex, blobSidecar, source, peerIdStr, seenTimestampSec}: AddBlob): void {
+    if (this.state.hasAllData) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INVALID_CONSTRUCTION,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot addBlob to BlockInputBlobs after it already is complete"
+      );
+    }
+
+    // this check suffices for checking slot, parentRoot, and forkName
+    if (blockRootHex !== this.blockRootHex) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISMATCHED_ROOT_HEX,
+          blockInputRoot: this.blockRootHex,
+          mismatchedRoot: blockRootHex,
+          source: source,
+          peerId: `${peerIdStr}`,
+        },
+        "Blob BeaconBlockHeader blockRootHex does not match BlockInput.blockRootHex"
+      );
+    }
+
+    if (this.state.hasBlock) {
+      assertBlockAndBlobArePaired(this.blockRootHex, this.state.block, blobSidecar);
+    }
+
+    // TODO: (@matthewkeil) check for duplicates and add metric here
+    // if (this.blobsCache.has(blobSidecar.index)) {
+    //   this.metrics.blockInput.duplicateBlob.inc()
+    // }
```

Comment:
Can remove this.  Now that all consumers will create BlockInput via the cache I've added the metrics and logging for this at that level.

https://github.com/ChainSafe/lodestar/blob/241b5badab13a08726a66e4f20a9651bfda2aa2b/packages/beacon-node/src/chain/seenCache/seenBlockInput.ts#L161-L175